File: src/main/java/io/github/ajoz/workshop/fp/java/part_3/solutions/solution_2/Exercise2.java
package io.github.ajoz.workshop.fp.java.part_3.solutions.solution_2;

import io.github.ajoz.workshop.fp.java.tools.Function2;
import io.github.ajoz.workshop.fp.java.tools.Supplier;

abstract class SealedBoolean {
    private SealedBoolean() {}

    public abstract SealedBoolean and(final SealedBoolean other);

    public abstract SealedBoolean or(final SealedBoolean other);

    public abstract SealedBoolean ifTrue(final Runnable effect);

    public abstract SealedBoolean ifFalse(final Runnable effect);

    public abstract <A> A match(final Supplier<A> ifTrue,
                                final Supplier<A> ifFalse);

    public static final SealedBoolean TRUE = new True();
    public static final SealedBoolean FALSE = new False();

    private static class True extends SealedBoolean {
        @Override
        public SealedBoolean and(final SealedBoolean other) {
            return other;
        }

        @Override
        public SealedBoolean or(final SealedBoolean other) {
            return this;
        }

        @Override
        public SealedBoolean ifTrue(final Runnable effect) {
            effect.run();
            return this;
        }

        @Override
        public SealedBoolean ifFalse(final Runnable effect) {
            return this;
        }

        @Override
        public <A> A match(final Supplier<A> ifTrue,
                           final Supplier<A> ifFalse) {
            return ifTrue.get();
        }

        @Override
        public String toString() {
            return "SealedTrue";
        }
    }

    private static class False extends SealedBoolean {

        @Override
        public SealedBoolean and(final SealedBoolean other) {
            return this;
        }

        @Override
        public SealedBoolean or(final SealedBoolean other) {
            return other;
        }

        @Override
        public SealedBoolean ifTrue(final Runnable effect) {
            return this;
        }

        @Override
        public SealedBoolean ifFalse(final Runnable effect) {
            effect.run();
            return this;
        }

        @Override
        public <A> A match(final Supplier<A> ifTrue,
                           final Supplier<A> ifFalse) {
            return ifFalse.get();
        }

        @Override
        public String toString() {
            return "SealedFalse";
        }
    }
}

public class Exercise2 {
    public static void main(String[] args) {
        System.out.println(SealedBoolean.TRUE.and(SealedBoolean.TRUE));
        System.out.println(SealedBoolean.TRUE.and(SealedBoolean.FALSE));
        System.out.println(SealedBoolean.FALSE.and(SealedBoolean.FALSE));
        System.out.println(SealedBoolean.TRUE.or(SealedBoolean.FALSE));
        System.out.println(SealedBoolean.FALSE.or(SealedBoolean.TRUE));
        System.out.println(SealedBoolean.FALSE.or(SealedBoolean.FALSE));

        SealedBoolean.TRUE
                .ifTrue(() -> System.out.println("SealedTrue is SealedTrue!"))
                .ifFalse(() -> System.out.println("SealedTrue is SealedFalse?"));

        SealedBoolean.FALSE
                .ifFalse(() -> System.out.println("SealedFalse is SealedFalse!"))
                .ifTrue(() -> System.out.println("SealedFalse is SealedTrue?"));

        final String trueMessage =
                SealedBoolean.TRUE
                        .match(
                                () -> "Matching SealedTrue to SealedTrue!",
                                () -> "Matching SealedTrue to SealedFalse?"
                        );

        System.out.println(trueMessage);

        final String falseMessage =
                SealedBoolean.FALSE
                        .match(
                                () -> "Matching SealedFalse to SealedTrue?",
                                () -> "Matching SealedFalse to SealedFalse!"
                        );

        System.out.println(falseMessage);
    }
}

File: kotlin/solutions/src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_3/solutions/exercise_2/Exercise2.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_3.solutions.exercise_2

import io.github.ajoz.workshop.fp.kotlin.tools.also

sealed class SealedBoolean {
    object True : SealedBoolean() {
        override fun and(other: SealedBoolean) = other
        override fun or(other: SealedBoolean) = this
        override operator fun not() = False

        override fun ifTrue(effect: () -> Unit) = also(effect)
        override fun ifFalse(effect: () -> Unit) = this

        override fun <A> match(ifTrue: () -> A, ifFalse: () -> A) = ifTrue()

        override fun toString(): String {
            return "SealedTrue"
        }
    }

    object False : SealedBoolean() {
        override fun and(other: SealedBoolean) = this
        override fun or(other: SealedBoolean) = other
        override operator fun not() = True

        override fun ifTrue(effect: () -> Unit) = this
        override fun ifFalse(effect: () -> Unit) = also(effect)

        override fun <A> match(ifTrue: () -> A, ifFalse: () -> A) = ifFalse()

        override fun toString(): String {
            return "SealedFalse"
        }
    }

    abstract infix fun and(other: SealedBoolean): SealedBoolean
    abstract infix fun or(other: SealedBoolean): SealedBoolean
    abstract operator fun not(): SealedBoolean

    abstract fun ifTrue(effect: () -> Unit): SealedBoolean
    abstract fun ifFalse(effect: () -> Unit): SealedBoolean

    abstract fun <A> match(ifTrue: () -> A, ifFalse: () -> A): A

    companion object {
        val TRUE: SealedBoolean = True
        val FALSE: SealedBoolean = False
    }
}

fun main(args: Array<String>) {
    // simple boolean algebra:
    // and:
    println(SealedBoolean.TRUE and SealedBoolean.TRUE)
    println(SealedBoolean.TRUE and SealedBoolean.FALSE)
    println(SealedBoolean.FALSE and SealedBoolean.FALSE)
    println(SealedBoolean.FALSE and SealedBoolean.TRUE)

    // or:
    println(SealedBoolean.TRUE or SealedBoolean.FALSE)
    println(SealedBoolean.FALSE or SealedBoolean.TRUE)
    println(SealedBoolean.FALSE or SealedBoolean.FALSE)
    println(SealedBoolean.TRUE or SealedBoolean.TRUE)

    // not:
    println(SealedBoolean.TRUE.not())
    println(SealedBoolean.FALSE.not())

    // conditions as methods:
    SealedBoolean.TRUE
            .ifTrue { println("SealedTrue is SealedTrue!") }
            .ifFalse { println("SealedTrue is SealedFalse?") }

    SealedBoolean.FALSE
            .ifFalse { println("SealedFalse is SealedFalse!") }
            .ifTrue { println("SealedFalse is SealedTrue?") }

    // conditions as expressions:
    val trueMessage = SealedBoolean.TRUE
            .match(
                    { "Matching SealedTrue to SealedTrue!" },
                    { "Matching SealedTrue to SealedFalse?" }
            )
    println(trueMessage)

    val falseMessage = SealedBoolean.FALSE
            .match(
                    { "Matching SealedFalse to SealedTrue?" },
                    { "Matching SealedFalse to SealedFalse!" }
            )
    println(falseMessage)

    val bool = SealedBoolean.FALSE
    when (bool) {
        SealedBoolean.FALSE -> println("Switch FALSE")
        SealedBoolean.TRUE -> println("Switch TRUE")
    }
}
